File: verdeciel/MainPanel.swift

//  Created by Devine Lu Linvega on 2015-09-21.
//  Copyright © 2015 XXIIVV. All rights reserved.

import UIKit
import QuartzCore
import SceneKit
import Foundation

class MainPanel : Panel
{
	let label = SCNLabel(text: "", scale: 0.1, align: alignment.center)
	let details = SCNLabel(text: "", scale: 0.085, align: alignment.center)
	var port:SCNPort!
	
	let header = Empty()
	let footer = Empty()
	
	var mainNode = Empty()
	var decals = Empty()
	
	override init()
	{
		super.init()
		
		name = "unknown"
		root.position = SCNVector3(x: 0, y: 0, z: templates.radius)
		root.addChildNode(mainNode)
		root.addChildNode(decals)
		
		// Header
		port = SCNPort(host: self)
		port.position = SCNVector3(x: 0, y: 0.4, z: templates.radius)
		label.position = SCNVector3(x: 0, y:0, z: templates.radius)
		header.addChildNode(port)
		header.addChildNode(label)
		addChildNode(header)
		header.eulerAngles.x += (degToRad(templates.titlesAngle))
		
		// Footer
		details.position = SCNVector3(x: 0, y: 0, z: templates.radius)
		footer.addChildNode(details)
		addChildNode(footer)
		footer.eulerAngles.x = (degToRad(-templates.titlesAngle))
		
		// Decals
		
		let width:CGFloat = 1.65
		let height:CGFloat = 1.8
		
		decals.addChildNode(SCNLine(vertices: [SCNVector3(width + 0.2,height - 0.2,0), SCNVector3(width,height,0), SCNVector3(width + 0.2,height - 0.2,0), SCNVector3(width + 0.2,-height + 0.2,0), SCNVector3(width + 0.2,-height + 0.2,0), SCNVector3(width,-height,0)],color:grey))
		decals.addChildNode(SCNLine(vertices: [SCNVector3(-width - 0.2,height - 0.2,0), SCNVector3(-width,height,0), SCNVector3(-width - 0.2,-height + 0.2,0), SCNVector3(-width,-height,0), SCNVector3(-width - 0.2,height - 0.2,0), SCNVector3(-width - 0.2,-height + 0.2,0)],color:grey))
		
		// Start
		
		mainNode.hide()
		decals.hide()
		footer.hide()
		
		label.update("--", color:grey)
	}
	
	override func whenStart()
	{
		super.whenStart()
		
		decals.hide()
		mainNode.hide()
		label.update("--", color: grey)
	}
	
	// MARK: Installation -
	
	var installNode:Empty = Empty()
	var installProgressBar = SCNProgressBar(width: 1)
	var installLabel = SCNLabel(text:"install", color:grey, align:.center)
	
	override func onInstallationBegin()
	{
		super.onInstallationBegin()
		
		helmet.addWarning("Installing", duration: 3, flag:"install")
		
		installNode = Empty()
		installNode.position = SCNVector3(0,0,0)
		installProgressBar = SCNProgressBar(width: 1)
		installProgressBar.position = SCNVector3(-installProgressBar.width/2,-0.3,0)
		installProgressBar.show()
		installNode.addChildNode(installProgressBar)
		
		installNode.addChildNode(installLabel)
		
		root.addChildNode(installNode)
	}
	
	override func installProgress()
	{
		super.installProgress()
		
		installLabel.update("Install \(Int(installPercentage))%")
		installProgressBar.update(installPercentage)
	}
	
	override func onInstallationComplete()
	{
		super.onInstallationComplete()
		
		mainNode.position = SCNVector3(0,0,-0.2)
		mainNode.hide()
		decals.position = SCNVector3(0,0,-0.4)
		decals.hide()
		
		SCNTransaction.begin()
		SCNTransaction.setAnimationDuration(0.7)
		mainNode.position = SCNVector3(0,0,0)
		mainNode.show()
		decals.position = SCNVector3(0,0,0)
		decals.show()
		footer.show()
		SCNTransaction.commit()
		
		installNode.removeFromParentNode()
		
		port.enable()
		label.update(name!, color: white)
	}
	
	required init?(coder aDecoder: NSCoder)
	{
		fatalError("init(coder:) has not been implemented")
	}
}

File: verreciel_js/scripts/panels/mainpanels/mainpanel.js
class MainPanel extends Panel
{
  constructor()
  {
    assertArgs(arguments, 0);
    super();

    this.installNode = new Empty();
    this.installProgressBar = new SceneProgressBar(1);
    this.installLabel = new SceneLabel("install", 0.1, Alignment.center, verreciel.grey);

    this.nameLabel = new SceneLabel("", 0.1, Alignment.center);
    this.detailsLabel = new SceneLabel("", 0.085, Alignment.center);
    this.header = new Empty();
    this.footer = new Empty();
    this.mainNode = new Empty();
    this.decals = new Empty();
    
    this.name = "unknown";
    this.details = "unknown";
    this.root.position.set(0, 0, Templates.radius);
    this.root.add(this.mainNode);
    this.root.add(this.decals);
    
    // Header
    this.port = new ScenePort(this);
    this.port.position.set(0, 0.4, Templates.radius);
    this.nameLabel.position.set(0, 0, Templates.radius);
    this.header.add(this.port);
    this.header.add(this.nameLabel);
    this.add(this.header);
    this.header.rotation.x += degToRad(Templates.titlesAngle);
    
    // Footer
    this.detailsLabel.position.set(0, 0, Templates.radius);
    this.footer.add(this.detailsLabel);
    this.add(this.footer);
    this.footer.rotation.x = degToRad(-Templates.titlesAngle);
    
    // Decals
    
    this.width = 1.65;
    this.height = 1.8;
    
    this.decals.add(new SceneLine([
      new THREE.Vector3(this.width + 0.2, this.height - 0.2, 0),
      new THREE.Vector3(this.width, this.height, 0),
      new THREE.Vector3(this.width + 0.2, this.height - 0.2, 0),
      new THREE.Vector3(this.width + 0.2, -this.height + 0.2, 0),
      new THREE.Vector3(this.width + 0.2, -this.height + 0.2, 0),
      new THREE.Vector3(this.width, -this.height, 0)
    ],
    verreciel.grey));
    this.decals.add(new SceneLine([
      new THREE.Vector3(-this.width - 0.2, this.height - 0.2, 0),
      new THREE.Vector3(-this.width, this.height, 0),
      new THREE.Vector3(-this.width - 0.2, -this.height + 0.2, 0),
      new THREE.Vector3(-this.width, -this.height, 0),
      new THREE.Vector3(-this.width - 0.2, this.height - 0.2, 0),
      new THREE.Vector3(-this.width - 0.2, -this.height + 0.2, 0)
    ],
    verreciel.grey));
    
    // Start
    
    this.mainNode.hide();
    this.decals.hide();
    this.footer.hide();
    
    this.nameLabel.updateText("--", verreciel.grey);
  }
  
  whenStart()
  {
    assertArgs(arguments, 0);
    super.whenStart();
    
    this.decals.hide();
    this.mainNode.hide();
    this.nameLabel.updateText("--", verreciel.grey);
  }
  
  // MARK: Installation -
  
  onInstallationBegin()
  {
    assertArgs(arguments, 0);
    super.onInstallationBegin();
    
    verreciel.helmet.addWarning("Installing", 3, "install");
    
    this.installNode = new Empty();
    this.installNode.position.set(0,0,0);
    this.installProgressBar = SceneProgressBar(1);
    this.installProgressBar.position.set(-this.installProgressBar.width/2,-0.3,0);
    this.installProgressBar.show();
    this.installNode.add(this.installProgressBar);
    
    this.installNode.add(this.installLabel);
    
    this.root.add(this.installNode);
  }
  
  installProgress()
  {
    assertArgs(arguments, 0);
    super.installProgress();
    
    this.installLabel.updateText("Install " + this.installPercentage.toFixed(0) + "%");
    this.installProgressBar.update(this.installPercentage);
  }
  
  onInstallationComplete()
  {
    assertArgs(arguments, 0);
    super.onInstallationComplete();
    
    this.mainNode.position.set(0,0,-0.2);
    this.mainNode.hide();
    this.decals.position.set(0,0,-0.4);
    this.decals.hide();
    
    // TODO: SCNTransaction

    // SCNTransaction.begin()
    // SCNTransaction.animationDuration = 0.7
    // this.mainNode.position.set(0,0,0);
    // this.mainNode.show();
    // this.decals.position.set(0,0,0);
    // this.decals.show();
    // footer.show()
    // SCNTransaction.commit()
    
    this.installNode.removeFromParentNode();
    
    this.port.enable();
    this.nameLabel.updateText(name, verreciel.white);
  }
}
